src/aspire/abinitio/commonline_cn.py

```diff
@@ -0,0 +1,376 @@
+
+        # Pre-compute shifted pf's.
+        pf_shifted = (pf * all_shift_phases[:, None, None]).swapaxes(0, 1)
+        pf_shifted = pf_shifted.reshape((n_img, n_shifts * (n_theta // 2), r_max))
+
+        # Step 1: pre-calculate the likelihood with respect to the self-common-lines.
+        scores_self_corrs = np.zeros((n_img, n_cands), dtype=self.dtype)
+        logger.info("Computing likelihood wrt self common-lines.")
+        for i in trange(n_img):
+            # Compute correlation of pf[i] with itself over all shifts.
+            corrs = pf_shifted[i] @ np.conj(pf_full[i]).T
+            corrs = np.reshape(corrs, (n_shifts, n_theta // 2, n_theta))
+            corrs_cands = np.max(
+                np.real(corrs[:, scls_inds[:, :, 0], scls_inds[:, :, 1]]), axis=0
+            )
+
+            scores_self_corrs[i] = np.mean(corrs_cands, axis=1)
+
+        # Step 2: Compute the likelihood for each pair of candidate matrices with respect
+        # to the common-lines they induce.
+        logger.info("Computing pairwise likelihood.")
+        n_vijs = n_img * (n_img - 1) // 2
+        vijs = np.zeros((n_vijs, 3, 3), dtype=self.dtype)
+        viis = np.zeros((n_img, 3, 3), dtype=self.dtype)
+        rots_symm = cyclic_rotations(self.order, self.dtype).matrices
+        c = 0
+
+        # List of MeanOuterProductEstimator instances.
+        # Used to keep a running mean of J-synchronized estimates for vii.
+        mean_est = []
+        for _ in range(n_img):
+            mean_est.append(MeanOuterProductEstimator())
+
+        with tqdm(total=n_vijs) as pbar:
+            for i in range(n_img):
+                for j in range(i + 1, n_img):
+                    # Compute correlation.
+                    corrs_ij = pf_shifted[i] @ np.conj(pf_full[j]).T
+
+                    # Max out over shifts.
+                    corrs_ij = np.max(
+                        np.reshape(
+                            np.real(corrs_ij), (n_shifts, n_theta // 2, n_theta)
+                        ),
+                        axis=0,
+                    )
+
+                    # Arrange correlation based on common lines induced by candidate rotations.
+                    corrs = corrs_ij[cijs_inds[..., 0], cijs_inds[..., 1]]
+                    corrs = np.reshape(
+                        corrs, (-1, self.order, n_theta_ijs // self.order)
+                    )
+                    # Take the mean over all symmetric common lines.
+                    corrs = np.mean(corrs, axis=1)
+                    corrs = np.reshape(
+                        corrs,
+                        (
+                            self.n_points_sphere,
+                            self.n_points_sphere,
+                            n_theta_ijs // self.order,
+                        ),
+                    )
+
+                    # Self common-lines are invariant to n_theta_ijs (i.e., in-plane rotation angles) so max them out.
```

Comment:
Can you elaborate on this? What is the index created below? Since we only need one of these `argmax`es, could this be delayed until later (L187)?

Reply:
I've added a note to explain a little better what is happening. 

What's happening here is we max over the "in-plane rotation" axis of `corrs` to compute the estimates for `viis`, which are invariant to in-plane rotations. But we have to keep around the index `opt_theta_ij_ind_per_sphere_points` of the best in-plane rotations for use in the estimate of `vijs`. I think we need to keep both `argmax`es around and in the same order (I could be missing something though).

Reply:
Ok, refactored this so we're only using a single `argmax` to get the indices of the optimal candidate rotations. Much more readable. Thank you!